libc/unistd_file.c

```diff
@@ -0,0 +1,848 @@
+#include <fcntl.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include <unity_fixture.h>
+
+#define LINE1 "line1\n"
+#define LINE2 "lineline2\n" /* longer than LINE1 */
+#define LINE3 "line3\n"     /* same length as LINE1 */
+#define LINE4 "\n"
+
+#define FNAME "unistd_file_testfile"
+
+
+static int fd, fd2;
+static char buf[50];
+
+
+/* Assert that string pointed to by str has been correctly written to a specified file */
+static void assert_write(const int fildes, const char *str)
+{
+	int toWrite;
+	int temp;
+	int cnt = 0;
+	int written = 0;
+
+	toWrite = strlen(str);
+	/* Assuming that strlen attempts would be sufficient */
+	while (written < toWrite && cnt++ <= toWrite) {
+		temp = write(fildes, &str[written], toWrite - written);
+		TEST_ASSERT_GREATER_OR_EQUAL(0, temp);
+		written += temp;
+	}
+
+	TEST_ASSERT_GREATER_OR_EQUAL(toWrite, written);
+}
+
+
+/* Assert that n characters have been read to dest from the specified file */
+static void assert_read(const int fildes, char *dest, const int n)
+{
+	int temp;
+	int cnt = 0;
+	int nread = 0;
+
+	/* Assuming that n attempts would be sufficient */
+	while (nread < n && cnt++ < n) {
+		temp = read(fildes, dest, n - nread);
+		TEST_ASSERT_GREATER_OR_EQUAL(0, temp);
+		nread += temp;
+	}
+
+	TEST_ASSERT_GREATER_OR_EQUAL(n, nread);
+}
+
+
+/* Assert that only n characters have been read to dest from the specified file.
+   Trying to read 2*n bytes in every attempt */
+static void assert_read_more(const int fildes, const char *dest, const int n)
+{
+	int temp;
+	int cnt = 0;
+	int nread = 0;
+
+	/* Assuming that n attempts would be sufficient */
+	while (nread < n && cnt++ < n) {
+		/* Always try to read 2 times more than it's possible */
+		temp = read(fildes, buf, 2 * n);
+		TEST_ASSERT_GREATER_OR_EQUAL(0, temp);
+		nread += temp;
+	}
+	TEST_ASSERT_EQUAL(n, nread);
+}
+
+
+TEST_GROUP(unistd_file);
+
+
+TEST_SETUP(unistd_file)
+{
+	/* open a file */
+	fd = open(FNAME, O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fd);
+	memset(buf, '\0', sizeof(buf));
+}
+
+
+TEST_TEAR_DOWN(unistd_file)
+{
+	memset(buf, '\0', sizeof(buf));
+	if (fd >= 0) {
+		TEST_ASSERT_EQUAL_INT(0, close(fd));
+	}
+	TEST_ASSERT_EQUAL_INT(0, remove(FNAME));
+}
+
+
+TEST(unistd_file, file_close)
+{
+	/* close an opened descriptor */
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+
+	/* try to close a closed descriptor */
+	TEST_ASSERT_EQUAL_INT(-1, close(fd));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	/* this test closes fd by itself, invalidate it */
+	fd = -1;
+}
+
+
+/* Simple usage of read() and write() */
+TEST(unistd_file, file_readwrite_nbytes)
+{
+	assert_write(fd, LINE1);
+
+	/* repoen file readonly */
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+	fd = open(FNAME, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fd);
+	{
+		/* try to read LINE1 from file */
+		assert_read(fd, buf, sizeof(LINE1) - 1);
+		TEST_ASSERT_EQUAL_STRING(LINE1, buf);
```

Comment:
Works by accident - `assert_write` has removed null terminator from LINE2, `buff` contains potentially unterminated string

Reply:
Sorry, stayed by accident, not true

Reply:
Yeah, but I will read and write including termination char now.